File: cli/evaluator.py
"""Texas Hold'em hand evaluation.

Given 5 to 7 cards, produce a comparable score so any two made hands can be
ranked. Higher score wins; equal scores tie (split pot).
"""

from __future__ import annotations

from collections import Counter
from dataclasses import dataclass

from cards import Card, RANK_TO_STR

# Hand category ranks, high is better.
HIGH_CARD = 0
ONE_PAIR = 1
TWO_PAIR = 2
THREE_KIND = 3
STRAIGHT = 4
FLUSH = 5
FULL_HOUSE = 6
FOUR_KIND = 7
STRAIGHT_FLUSH = 8

CATEGORY_NAME = {
    HIGH_CARD: "High Card",
    ONE_PAIR: "Pair",
    TWO_PAIR: "Two Pair",
    THREE_KIND: "Three of a Kind",
    STRAIGHT: "Straight",
    FLUSH: "Flush",
    FULL_HOUSE: "Full House",
    FOUR_KIND: "Four of a Kind",
    STRAIGHT_FLUSH: "Straight Flush",
}

_PLURAL = {
    2: "Twos", 3: "Threes", 4: "Fours", 5: "Fives", 6: "Sixes", 7: "Sevens",
    8: "Eights", 9: "Nines", 10: "Tens", 11: "Jacks", 12: "Queens",
    13: "Kings", 14: "Aces",
}


@dataclass(frozen=True, order=True)
class HandResult:
    key: tuple  # comparable; (category, tiebreak1..tiebreak5)
    category: int
    name: str


def _best_straight_high(rank_set: set[int]) -> int | None:
    """Highest card of the best straight in rank_set, or None.

    Treats Ace (14) as also low (1) so the wheel A-2-3-4-5 is detected.
    """
    s = set(rank_set)
    if 14 in s:
        s.add(1)
    for high in range(14, 4, -1):
        if all((high - i) in s for i in range(5)):
            return high
    return None


def _name_for(category: int, tb: list[int]) -> str:
    if category == STRAIGHT_FLUSH:
        if tb[0] == 14:
            return "Royal Flush"
        return f"Straight Flush, {RANK_TO_STR[tb[0]]} high"
    if category == FOUR_KIND:
        return f"Four of a Kind, {_PLURAL[tb[0]]}"
    if category == FULL_HOUSE:
        return f"Full House, {_PLURAL[tb[0]]} over {_PLURAL[tb[1]]}"
    if category == FLUSH:
        return f"Flush, {RANK_TO_STR[tb[0]]} high"
    if category == STRAIGHT:
        return f"Straight, {RANK_TO_STR[tb[0]]} high"
    if category == THREE_KIND:
        return f"Three of a Kind, {_PLURAL[tb[0]]}"
    if category == TWO_PAIR:
        return f"Two Pair, {_PLURAL[tb[0]]} and {_PLURAL[tb[1]]}"
    if category == ONE_PAIR:
        return f"Pair of {_PLURAL[tb[0]]}"
    return f"High Card {RANK_TO_STR[tb[0]]}"


def evaluate(cards: list[Card]) -> HandResult:
    """Evaluate the best 5-card hand from 5..7 cards."""
    if len(cards) < 5:
        raise ValueError("Need at least 5 cards to evaluate.")

    ranks = [c.rank for c in cards]
    rank_counts = Counter(ranks)
    # Sort (rank, count) by count desc, then rank desc.
    by_count = sorted(rank_counts.items(), key=lambda kv: (kv[1], kv[0]), reverse=True)
    distinct_ranks_desc = sorted(rank_counts.keys(), reverse=True)

    # Flush detection: a suit with >= 5 cards.
    suit_counts = Counter(c.suit for c in cards)
    flush_suit = next((s for s, n in suit_counts.items() if n >= 5), None)
    flush_ranks: list[int] = []
    if flush_suit is not None:
        flush_ranks = sorted((c.rank for c in cards if c.suit == flush_suit), reverse=True)

    # Straight flush: a straight inside the flush suit.
    if flush_suit is not None:
        sf_high = _best_straight_high(set(flush_ranks))
        if sf_high is not None:
            tb = [sf_high]
            return HandResult((STRAIGHT_FLUSH, *_pad(tb)), STRAIGHT_FLUSH, _name_for(STRAIGHT_FLUSH, tb))

    # Four of a kind.
    if by_count[0][1] == 4:
        quad = by_count[0][0]
        kicker = max(r for r in distinct_ranks_desc if r != quad)
        tb = [quad, kicker]
        return HandResult((FOUR_KIND, *_pad(tb)), FOUR_KIND, _name_for(FOUR_KIND, tb))

    # Full house (trips + a separate pair-or-better).
    trips = [r for r, c in by_count if c >= 3]
    pairs = [r for r, c in by_count if c >= 2]
    if trips:
        trip_rank = trips[0]
        pair_rank = next((r for r in pairs if r != trip_rank), None)
        if pair_rank is not None:
            tb = [trip_rank, pair_rank]
            return HandResult((FULL_HOUSE, *_pad(tb)), FULL_HOUSE, _name_for(FULL_HOUSE, tb))

    # Flush.
    if flush_suit is not None:
        tb = flush_ranks[:5]
        return HandResult((FLUSH, *_pad(tb)), FLUSH, _name_for(FLUSH, tb))

    # Straight.
    straight_high = _best_straight_high(set(distinct_ranks_desc))
    if straight_high is not None:
        tb = [straight_high]
        return HandResult((STRAIGHT, *_pad(tb)), STRAIGHT, _name_for(STRAIGHT, tb))

    # Three of a kind.
    if trips:
        trip_rank = trips[0]
        kickers = [r for r in distinct_ranks_desc if r != trip_rank][:2]
        tb = [trip_rank, *kickers]
        return HandResult((THREE_KIND, *_pad(tb)), THREE_KIND, _name_for(THREE_KIND, tb))

    # Two pair.
    if len(pairs) >= 2:
        high_pair, low_pair = pairs[0], pairs[1]
        kicker = max(r for r in distinct_ranks_desc if r not in (high_pair, low_pair))
        tb = [high_pair, low_pair, kicker]
        return HandResult((TWO_PAIR, *_pad(tb)), TWO_PAIR, _name_for(TWO_PAIR, tb))

    # One pair.
    if len(pairs) == 1:
        pair_rank = pairs[0]
        kickers = [r for r in distinct_ranks_desc if r != pair_rank][:3]
        tb = [pair_rank, *kickers]
        return HandResult((ONE_PAIR, *_pad(tb)), ONE_PAIR, _name_for(ONE_PAIR, tb))

    # High card.
    tb = distinct_ranks_desc[:5]
    return HandResult((HIGH_CARD, *_pad(tb)), HIGH_CARD, _name_for(HIGH_CARD, tb))


def _pad(tiebreak: list[int]) -> list[int]:
    """Pad a tiebreak list to exactly 5 entries so all keys compare cleanly."""
    return (tiebreak + [0, 0, 0, 0, 0])[:5]

File: web/src/evaluator.js
// Texas Hold'em hand evaluation. Mirrors the Python evaluator.
// Produces an integer score (higher = better); equal scores tie.

import { RANK_STR } from "./cards.js";

export const HIGH_CARD = 0, ONE_PAIR = 1, TWO_PAIR = 2, THREE_KIND = 3,
  STRAIGHT = 4, FLUSH = 5, FULL_HOUSE = 6, FOUR_KIND = 7, STRAIGHT_FLUSH = 8;

export const CATEGORY_NAME = {
  0: "High Card", 1: "Pair", 2: "Two Pair", 3: "Three of a Kind",
  4: "Straight", 5: "Flush", 6: "Full House", 7: "Four of a Kind",
  8: "Straight Flush",
};

const PLURAL = {
  2: "Twos", 3: "Threes", 4: "Fours", 5: "Fives", 6: "Sixes", 7: "Sevens",
  8: "Eights", 9: "Nines", 10: "Tens", 11: "Jacks", 12: "Queens",
  13: "Kings", 14: "Aces",
};

function bestStraightHigh(rankSet) {
  const s = new Set(rankSet);
  if (s.has(14)) s.add(1); // wheel: A plays low
  for (let high = 14; high >= 5; high--) {
    let ok = true;
    for (let i = 0; i < 5; i++) if (!s.has(high - i)) { ok = false; break; }
    if (ok) return high;
  }
  return null;
}

// Encode category + up to 5 tiebreakers (each < 16) into one comparable int.
function encode(category, tb) {
  const t = tb.concat([0, 0, 0, 0, 0]).slice(0, 5);
  return ((((category * 16 + t[0]) * 16 + t[1]) * 16 + t[2]) * 16 + t[3]) * 16 + t[4];
}

function nameFor(category, tb) {
  switch (category) {
    case STRAIGHT_FLUSH: return tb[0] === 14 ? "Royal Flush" : `Straight Flush, ${RANK_STR[tb[0]]} high`;
    case FOUR_KIND: return `Four of a Kind, ${PLURAL[tb[0]]}`;
    case FULL_HOUSE: return `Full House, ${PLURAL[tb[0]]} over ${PLURAL[tb[1]]}`;
    case FLUSH: return `Flush, ${RANK_STR[tb[0]]} high`;
    case STRAIGHT: return `Straight, ${RANK_STR[tb[0]]} high`;
    case THREE_KIND: return `Three of a Kind, ${PLURAL[tb[0]]}`;
    case TWO_PAIR: return `Two Pair, ${PLURAL[tb[0]]} and ${PLURAL[tb[1]]}`;
    case ONE_PAIR: return `Pair of ${PLURAL[tb[0]]}`;
    default: return `High Card ${RANK_STR[tb[0]]}`;
  }
}

function result(category, tb) {
  return { score: encode(category, tb), category, name: nameFor(category, tb) };
}

// cards: array of Card (5..7). Returns { score, category, name }.
export function evaluate(cards) {
  const rankCounts = new Map();
  for (const c of cards) rankCounts.set(c.rank, (rankCounts.get(c.rank) || 0) + 1);

  // (rank, count) sorted by count desc, then rank desc.
  const byCount = [...rankCounts.entries()].sort((a, b) =>
    b[1] - a[1] || b[0] - a[0]);
  const distinctDesc = [...rankCounts.keys()].sort((a, b) => b - a);

  // Flush.
  const suitCounts = new Map();
  for (const c of cards) suitCounts.set(c.suit, (suitCounts.get(c.suit) || 0) + 1);
  let flushSuit = null;
  for (const [s, n] of suitCounts) if (n >= 5) { flushSuit = s; break; }
  let flushRanks = [];
  if (flushSuit !== null) {
    flushRanks = cards.filter(c => c.suit === flushSuit).map(c => c.rank).sort((a, b) => b - a);
  }

  // Straight flush.
  if (flushSuit !== null) {
    const sfHigh = bestStraightHigh(flushRanks);
    if (sfHigh !== null) return result(STRAIGHT_FLUSH, [sfHigh]);
  }

  // Four of a kind.
  if (byCount[0][1] === 4) {
    const quad = byCount[0][0];
    const kicker = Math.max(...distinctDesc.filter(r => r !== quad));
    return result(FOUR_KIND, [quad, kicker]);
  }

  // Full house.
  const trips = byCount.filter(([, c]) => c >= 3).map(([r]) => r);
  const pairs = byCount.filter(([, c]) => c >= 2).map(([r]) => r);
  if (trips.length) {
    const tripRank = trips[0];
    const pairRank = pairs.find(r => r !== tripRank);
    if (pairRank !== undefined) return result(FULL_HOUSE, [tripRank, pairRank]);
  }

  // Flush.
  if (flushSuit !== null) return result(FLUSH, flushRanks.slice(0, 5));

  // Straight.
  const straightHigh = bestStraightHigh(distinctDesc);
  if (straightHigh !== null) return result(STRAIGHT, [straightHigh]);

  // Three of a kind.
  if (trips.length) {
    const tripRank = trips[0];
    const kickers = distinctDesc.filter(r => r !== tripRank).slice(0, 2);
    return result(THREE_KIND, [tripRank, ...kickers]);
  }

  // Two pair.
  if (pairs.length >= 2) {
    const [hi, lo] = pairs;
    const kicker = Math.max(...distinctDesc.filter(r => r !== hi && r !== lo));
    return result(TWO_PAIR, [hi, lo, kicker]);
  }

  // One pair.
  if (pairs.length === 1) {
    const pr = pairs[0];
    const kickers = distinctDesc.filter(r => r !== pr).slice(0, 3);
    return result(ONE_PAIR, [pr, ...kickers]);
  }

  // High card.
  return result(HIGH_CARD, distinctDesc.slice(0, 5));
}
